satellite/gc/piecetracker.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+// PieceTracker implements the metainfo loop observer interface for garbage collection
+type PieceTracker struct {
+	log          *zap.Logger
+	config       Config
+	creationDate time.Time
+	pieceCounts  map[storj.NodeID]int
+
+	retainInfos map[storj.NodeID]*RetainInfo
+}
+
+// NewPieceTracker instantiates a new gc piece tracker to be subscribed to the metainfo loop
+func NewPieceTracker(log *zap.Logger, config Config, pieceCounts map[storj.NodeID]int) *PieceTracker {
+	return &PieceTracker{
+		log:          log,
+		config:       config,
+		creationDate: time.Now().UTC(),
+		pieceCounts:  pieceCounts,
+
+		retainInfos: make(map[storj.NodeID]*RetainInfo),
+	}
+}
+
+// RemoteSegment takes a remote segment found in metainfo and adds pieces to bloom filters
+func (pieceTracker *PieceTracker) RemoteSegment(ctx context.Context, path storj.Path, pointer *pb.Pointer) (err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
Wouldn't be good to trace the `path` passing it to the Task function (i.e. `defer mon.Task()(&ctx, path)(&err)`)